storage/innobase/row/row0ins.cc

```diff
@@ -1867,6 +1870,21 @@ row_ins_check_foreign_constraint(
 			&check_table->n_foreign_key_checks_running, 1);
 
 		lock_wait_suspend_thread(thr);
+#ifdef WITH_WSREP
+		switch (trx->error_state) {
+		case DB_DEADLOCK:
+			if (wsrep_debug) {
+				ib::info() <<
+				"WSREP: innodb trx state changed during wait "
+				<< " trx: " << trx->id << " with error_state: "
+				<< trx->error_state << " err: " << err;
+			}
+			err = trx->error_state;
+			break;
+		default:
+			break;
+		}
+#endif /* WITH_WSREP */
 
 		thr->lock_state = QUE_THR_LOCK_NOLOCK;
```

Comment:
It looks like this was fixed later in commit a21e01a53d309c4c949d41f85a43211008bac1fc with the following message:
>MDEV-17541 KILL QUERY during lock wait in FOREIGN KEY check causes hang
>row_ins_check_foreign_constraint(): Do not overwrite hard errors with the soft error DB_LOCK_WAIT. This prevents an infinite wait loop when DB_INTERRUPTED was returned. For DB_LOCK_WAIT, row_insert_for_mysql() would keep invoking row_ins_step() and the transaction would remain active until the server shutdown is initiated.

Side note: I think that the debug message is somewhat redundant. The values of `err` and `trx->error_state` are known at that point: `DB_LOCK_WAIT` and `DB_DEADLOCK`. Was it intentional to read `trx->error_state` 3 times?

Reply:
The fix in this PR protects that error coder overwrite does not happen for DB_DEADLOCK error, which was result of high priority replication abort.
The fix for MDEV-17541, otoh is more general and is sort of superclass of this fix.

However, the mtr test in this PR would be good to have in the product.